File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> solution(vector<int> numbers) {
    vector<int> answer(numbers.size(), -1);
    stack<int> s;

    s.push(0);

    for(int i = 1; i < numbers.size(); i++){
        int current = numbers[i];
        while(!s.empty()){
            if(numbers[s.top()] < current){
                answer[s.top()] = current;
                s.pop();
            }else break;
        }
        s.push(i);
    }
   
    return answer;
}

int main(){
    solution({2,3,3,5});
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.js
function solution(numbers) {
  var answer = [];

  var stack = [];

  for(let i = 0; i<numbers.length; i++){
    answer.push(-1);
  }
  for(let i = numbers.length - 1; i >= 0; i--){
    while(stack.length && stack[0] <= numbers[i]) stack.shift();

    if(stack.length){
      answer[i] = stack[0];
    }

    stack.unshift(numbers[i]);
  }

  return answer;
}